keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,406 @@
+* This proposal only target CSI volumes.
+
+### Non-Goals
+
+* Node/control plane partitioning other than a node shutdown is not covered by the proposal, but will be addressed in the future.
+* Implement in-cluster logic to handle node/control plane partitioning. 
+* Enable detach for all the storage providers.
+* Existing in-tree volumes are not targeted by this proposal.
+
+## Proposal
+
+User stories:
+
+* As a cluster administrator I want my stateful workload to failover in case a node is shutdown without any intervention when it's safe to do so.
+
+* As a developer I would like to rely on a self-healing platform for my stateful workloads.
+
+* In the [CSIDriverSpec](https://github.com/kubernetes/kubernetes/blob/v1.20.2/pkg/apis/storage/types.go#L266), add a `SafeDetach` boolean to indicate if it is safe to detach.
+
+```go
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+  ...
+
+  // SafeDetach indicates this CSI volume driver is able to perform detach
+  // operations. When set to true, the CSI volume driver needs to ensure that 
+  // ControllerUnpublishVolume() implementation is checking whether a volume
+  // can be detached safely
+  // +optional
+  SafeDetach *bool
+}
+```
+
+* The Graceful Node Shutdown [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown) introduced a way to detect a node shutdown. When the NodeShutdownManager detects a node is shutting down, it sets the `Reason` field of a `PodStatus` to a constant named `nodeShutdownReason`. Currently `nodeShutdownReason` is in the `nodeshutdown` package. We need to export this constant by renaming it to `NodeShutdownReason` and move it to a package that can be imported by the Pod GC Controller.
+
+Existing logic:
+* When a node is partitioned from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.
+
+* After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.
+
+Proposed logic change:
+* The Pod GC Controller, part of the kube-controller-manager, would need to go through all the Pods in terminating state, verify that the Node is NotReady, check if the `Reason` in the `PodStatus` is set to the `NodeShutDownReason` constant, and then check if the `SafeDetach` flag in CSIDriver is set to true. If so, forcefully delete pods that are stuck.
+
+* Specifically the Pod GC Controller (the only controller able to forcefully delete pods) would also check now for stateful pods, and select those that satisfy ALL of the following conditions:
+
+  * pods that are backed by a CSI driver that supports `SafeDetach`.
+  * pods that are backed by volumes that aren't marked as `ReadWriteMany`.
+
```

Comment:
why `ReadWriteMany` is forbidden?

Reply:
I think that we can already attach a `ReadWriteMany` volume to another pods in other nodes, even when the node that the volume is used on is unresponsive. Therefore, there will be no need for extra special handlings in this case.

Reply:
I'm not sure why we need to explicitly treat RWX specially if a driver says it can support safe detach?

Reply:
Removed this line.